Autorename

taniwha opened the pull request:
This allows KIS items to specify they want to use Vessel.AutoRename based on the originating vessel's name rather than the item part's title. The default is to use the item part's title, but setting vesselAutoRename to true in the MODULE block alters this. This is most useful for EL's survey stakes, but other mods may find it useful.

taniwha committed:
Update Makefile.

taniwha committed:
Ignore the build directory.

taniwha committed:
Clean up compile warnings.

taniwha committed:
Allow KIS items to use Vessel.AutoRename.

Sometimes KSP's automatic vessel naming is useful (particularly for the
survey stakes in Extraplanetary Launchpads). The default is still to use
the part's title, but if ModuleKISItem.vesselAutoRename is true, then the
name will instead be taken from the originating inventory part.

taniwha committed:
Better config and pod inventory setup.

KISAddonConfig now runs during the loading scene, first loading the configs
somewhere between ModuleManager and PartLoader, and then adding pod
inventories sometime after PartLoader (other mods might insert their
loaders in the intervals). The goal is to allow KIS-aware mods to add pod
inventories to parts that have dynamic crew capacities.

ihsoft commented on Source/KIS_Shared.cs:
```diff
@@ -1363,7 +1363,13 @@ public enum MessageAction {
     if (newVessel.Initialize(false)) {
-      newVessel.vesselName = newPart.partInfo.title;
+      var item = newPart.FindModuleImplementing <ModuleKISItem> ();
+      if (item != null && !item.vesselAutoRename) {
```
This condition doesn't keep the default behavior. In case of there is no module `ModuleKISItem`, then new naming mechanism will be used.

taniwha replied:
oops, inverted logic.

ihsoft commented on Source/KIS_Shared.cs:
```diff
@@ -1354,7 +1354,7 @@ public enum MessageAction {
   /// <summary>Creates a new vessel from a single part.</summary>
   /// <remarks>Initially the part must belong to some vessel.</remarks>
-  static IEnumerator WaitAndMakeLonePart(Part newPart, OnPartReady onPartReady) {
+  static IEnumerator WaitAndMakeLonePart(Part newPart, Part fromPart, OnPartReady onPartReady) {
```
You don't need an extra parameter. Initially, `newPart` belongs to the from vessel. Simply save the vessel (or the name) at the beginning of the method.

taniwha replied:
Ah, ok. I just tested and that does indeed work nicely. Have a problem with ground attaching, though (probably something went wrong in the config patch).

taniwha replied:
Turns out the attach problem was PEBCAK: I was using the drill instead of EL's mallet.

taniwha committed:
Avoid adding duplicate pod inventories.

If some other mod has added pod inventories to the part, skip the part.

ihsoft reviewed (changes requested):
See comments. Also, I see a lot of other unrelated changes. Is it your working version?

ihsoft commented on Source/KISAddonConfig.cs:
```diff
@@ -22,52 +23,133 @@ sealed class KISAddonConfig : MonoBehaviour {
+          DebugEx.Fine("Found part with CrewCapacity: {0}", avPart.name);
+          AddPodInventories (avPart.partPrefab, avPart.partPrefab.CrewCapacity);
+        }
+        yield return null;
```
Returning `null` will result in loading one part per a frame. On the low-end machines with many addons it may take forever to complete. This is definitely not the case when smooth loading is better than quick loading.  At least do some batching here. And, in general, let's extract this refactoring into a separate PR. I'd like to test it before saying "yes".

ihsoft commented on Source/KISAddonConfig.cs:
```diff
@@ -22,52 +23,133 @@ sealed class KISAddonConfig : MonoBehaviour {
+    List<LoadingSystem> list = LoadingScreen.Instance.loaders;
+    if (list != null) {
+      for (int i = 0; i < list.Count; i++) {
+        if (list[i] is KISConfigLoader) {
```
Loading the configs is a cheap operation. Just load it in KISPodInventoryLoader.

taniwha replied:
See overall reply, but I realized I did add comments.

taniwha commented:
It's not about expense, but allowing other mods to call AddPodInventories in their OnLoad methods. ie, configs need to be loaded before parts so part modules can (indirectly) reference them.

I'll add some comments.

taniwha commented:
Most of the other changes (makefile, warnings) were just to get a build (using make and -warnaserror). The inventory set would up on the same PR due to not being sure what to do with it.

taniwha committed:
Correct inverted logic on autorename test.

Gets me more times that I like.

taniwha committed:
Cleaner handling of vessel name.

It turns out newPart's vessel is initially the originating vessel, so this
works nicely (thanks, igorz).

ihsoft commented:
Can we split the PR into two (or more) PRs? I'm OK with the auto rename thing. However, the mod loading refactoring is not that obvious to me at this point. It would be better to discuss/review it separately.

taniwha commented:
Sure, I'll sort it out in the next day or two.

ihsoft commented:
Thanks.

Btw, take a look at the [coding style definitions](https://github.com/ihsoft/KIS/blob/master/Source/README.md). I'm not a "grammar-nazi", but given how much time I spend working with the code, I'd prefer it to be following some syntax rules :)

taniwha commented:
Closing in favor of newer PRs.
